_overviews/scala3-book/control-structures.md

```diff
@@ -989,4 +990,178 @@ finally
+      for j <- 1 to 10 do
+        if i == 7 && j == 9 then break("found")
+    "not found"
+
+println(result)
+```
+
+{% endtab %}
+{% endtabs %}
+
+Since `break` uses an exception to control the execution flow, mixing it with `try`/`catch` can lead to unexpected results:
+
+{% tabs control-structures-33 %}
+{% tab 'Scala 3 Only'  %}
+
+```scala
+import scala.util.boundary, boundary.break
+
+val result = boundary:
+    try
+      for i <- 1 to 10 do
+        some_operation(i)
+        if i == 5 then break(i)
+    catch
+      case e: Exception => -1
+
+println(result) // prints -1, not 5
+```
+
+{% endtab %}
+{% endtabs %}
+
+The right way to use it would be to place the `boundary` expression inside the `try` block:
+
+{% tabs control-structures-34 %}
+{% tab 'Scala 3 Only'  %}
+
+```scala
+import scala.util.boundary, boundary.break
+
+val result =
+    try
+      boundary:
+        for i <- 1 to 10 do
+          some_operation(i)
+          if i == 5 then break(i)
+    catch
+      case e: Exception => -1
+
+println(result) // 5
+```
+
+{% endtab %}
+{% endtabs %}
+
+## Custom control structures
+
+The use of Scala 3 quiet syntax, by-name parameters, context parameters  and `boundary`/`break` allows for writing code
+that resembles the built-in control structures and helps with reducing the boilerplate.
+In this example, the `repeat` function wraps a simple `for` loop, and the by-name `action` parameter will be executed `n`
+times for its side effects.
+
+{% tabs control-structures-35 %}
+{% tab 'Scala 3 Only'  %}
+
+```scala
+def repeat(n: Int)(action: => Unit): Unit =
+  for i <- 1 to n do
+    action
+
+repeat(5):
+  println("test")
+```
+
+{% endtab %}
+{% endtabs %}
+
+Here's an example that simulates a `do`-`while` loop by using `boundary`/`break`:
+
+{% tabs control-structures-36 %}
+{% tab 'Scala 3 Only'  %}
+
+```scala
+import scala.util.boundary, boundary.break, boundary.Label
+
+def loop(action: Label[Unit] ?=> Unit): Unit =
+  boundary[Unit]:
+    while true do
+      action
+
+def until(condition: => Boolean)(using Label[Unit]): Unit =
+  if condition then break()
+```
+
+{% endtab %}
+{% endtabs %}
+
+The `loop` function will continue to execute `action` indefinitely, unless the execution breaks out of the internal 
+`while` loop. The `until` function will stop the execution of `loop` if `condition` is met. `break()` needs a context parameter
+of type `Label[Unit]`, denoting which `boundary` it is supposed to break to.
+To use the `loop`/`until` structure, pass the code to be executed as an argument to `loop`, including the `until` check:
+
+{% tabs control-structures-37 %}
+{% tab 'Scala 3 Only'  %}
+
+```scala
+var i = 0
+loop:
+  println(i)
+  i = i + 1
+  until(i == 10)
+```
+
+{% endtab %}
+{% endtabs %}
+
+The following example uses a context parameter to store actions to be executed after a block of code is completed. The
+`with_defer` function defines a scope in which `defer` can be used to defer execution of actions.
```

Comment:
Not sure about this because context parameters are not mentioned yet at this point. Maybe the custom control structures section should be moved to a separate page?